stored_requests/backends/db_fetcher/fetcher.go

```diff
@@ -5,28 +5,63 @@ import (
 	"database/sql"
 	"encoding/json"
 	"fmt"
+
 	"github.com/golang/glog"
+	"github.com/prebid/prebid-server/stored_requests/cache/cacher"
 )
 
 // dbFetcher fetches Stored Requests from a database. This should be instantiated through the NewPostgres() function.
 type dbFetcher struct {
 	db         *sql.DB
 	queryMaker func(int) (string, error)
+	cache      cacher.Cacher
```

Comment:
This can be much more composable. I'd recommend that the core Cache code should live in a [Decorator](https://en.wikipedia.org/wiki/Decorator_pattern). For example something like this:

```
func AddCache(cache Cacher, delegate Fetcher) Fetcher {
  // return a Fetcher which checks the cache, but uses the delegate if the cache misses
}
```

Benefits being:

- Any cache layers will work with any backend. They won't be limited to DBs.
- Can define multiple cache layers. For example, you may want a small in-memory cache, a larger redis cache, and then the DB as a final fallback.
- Easier to configure caches through Viper (see my comment below)
- The smaller pieces will be easier to test. Mocking out the DB is a pain... but with the above, you use a `map`-backed Fetcher in unit tests to test your Cache code.

Reply:
How would it look if there are multiple cache layers? Would one first layer have a shorter TTL than the second cache?

Reply:
> How would it look if there are multiple cache layers?

I was imagining something like this:

```
func makeFetcher(cfg *config.Configuration) (fetcher Fetcher) {
  fetcher = makeBaseFetcher(cfg.fetcher.base)
  for _, cacheConfig := cfg.fetcher.cacheLayers {
    fetcher = AddCache(makeCacher(cacheConfig, fetcher)
  }
  return
}

func makeCacher(cfg *config.CacheConfig) Cacher {
  ttl := cfg.ttl
  if isMemory(cfg) {
    return &inmemorycache.Cache{
      ttl: cfg.ttl,
    }
  } else if isRedis(cfg.ttl) ...
}
```

This uses a pretty crappy config setup... but gets the general idea across.

> Would one first layer have a shorter TTL than the second cache?

In practice... maybe? Probably? That'd be up to whoever is hosting PBS. Each type of cache would have its own configuration options... so I assume they'd use either TTLs, or max capacities, or things like that to optimize for their hardware.

Reply:
In this code it looks like there is one cache used per deployment?

Reply:
akh... my bad. That should read:

```
for _, cacheConfig := range cfg.fetcher.cacheLayers {
    fetcher = AddCache(makeCacher(cacheConfig, fetcher)
  }
```

that's what I get for skipping the IDE and coding directly in github 🙄

Reply:
But it's still an either/or due to the if/else? 

```
if isMemory(cfg) {
    return &inmemorycache.Cache{
      ttl: cfg.ttl,
    }
  } else if isRedis(cfg.ttl) ..
```

Reply:
A _single_ cache layer must be either/or. A _list_ of cache layers may include one of each type.

Reply:
That said... I haven't actually checked to see if Viper supports struct slices yet. If not, that could blow a hole in this whole idea.

Reply:
I believe Viper supports maps but I don't think it supports slice of structs (just slice of strings)

Reply:
> Can define multiple cache layers. For example, you may want a small in-memory cache, a larger redis cache, and then the DB as a final fallback.

using this example... What happens if we have a miss in the in-memory cache and a hit in redis? Do we set the value in the smaller cache or do we only update that cache when we hit the DB/postgres?